File: TypeScript/clbg/reverse-complement/revcomp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const n = Number.parseInt(process.argv[2], 10);
void n;
const data = fs.readFileSync(0);
const complement = Buffer.allocUnsafe(256);
for (let i = 0; i < 256; i++) {
    complement[i] = i;
}
const source = "ACBDGHKMNSRUTWVYacbdghkmnsrutwvy";
const target = "TGVHCDMKNSYAAWBRTGVHCDMKNSYAAWBR";
for (let i = 0; i < source.length; i++) {
    complement[source.charCodeAt(i)] = target.charCodeAt(i);
}
function reverseComplement(start, end) {
    let left = start;
    let right = end - 1;
    while (left <= right) {
        while (left <= right && (data[left] === 10 || data[left] === 13)) {
            left++;
        }
        while (left <= right && (data[right] === 10 || data[right] === 13)) {
            right--;
        }
        if (left > right) {
            break;
        }
        if (left === right) {
            data[left] = complement[data[left]];
            break;
        }
        const leftBase = data[left];
        const rightBase = data[right];
        data[left] = complement[rightBase];
        data[right] = complement[leftBase];
        left++;
        right--;
    }
}
let sequenceStart = -1;
let position = 0;
while (position < data.length) {
    if (data[position] === 62 &&
        (position === 0 || data[position - 1] === 10)) {
        if (sequenceStart >= 0) {
            reverseComplement(sequenceStart, position);
        }
        const newline = data.indexOf(10, position);
        if (newline === -1) {
            sequenceStart = data.length;
            position = data.length;
            break;
        }
        sequenceStart = newline + 1;
        position = sequenceStart;
    }
    else {
        position++;
    }
}
if (sequenceStart >= 0 && sequenceStart < data.length) {
    reverseComplement(sequenceStart, data.length);
}
process.stdout.write(data);

File: Java/clbg/reverse-complement/revcomp.java
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;

public class revcomp {
    private static final byte[] COMPLEMENT = new byte[256];

    static {
        for (int i = 0; i < COMPLEMENT.length; i++) {
            COMPLEMENT[i] = (byte) i;
        }

        String bases = "ACGTUMRWSYKVHDBN";
        String complements = "TGCAAKYWSRMBDHVN";
        for (int i = 0; i < bases.length(); i++) {
            char base = bases.charAt(i);
            byte complement = (byte) complements.charAt(i);
            COMPLEMENT[base] = complement;
            COMPLEMENT[Character.toLowerCase(base)] = complement;
        }
    }

    private static void reverseComplement(byte[] data, int start, int end) {
        int left = start;
        int right = end - 1;

        while (left <= right) {
            byte a = data[left];
            if (a == '\n' || a == '\r') {
                left++;
                continue;
            }

            byte b = data[right];
            if (b == '\n' || b == '\r') {
                right--;
                continue;
            }

            data[left++] = COMPLEMENT[b & 255];
            data[right--] = COMPLEMENT[a & 255];
        }
    }

    public static void main(String[] args) throws Exception {
        int n = Integer.parseInt(args[0]);

        InputStream input = System.in;
        byte[] data = new byte[65536];
        int length = 0;

        while (true) {
            if (length == data.length) {
                int capacity = data.length <= (Integer.MAX_VALUE - 8) / 2
                        ? data.length * 2
                        : Integer.MAX_VALUE - 8;
                if (capacity <= data.length) {
                    throw new OutOfMemoryError("FASTA input is too large");
                }
                data = Arrays.copyOf(data, capacity);
            }

            int count = input.read(data, length, data.length - length);
            if (count < 0) {
                break;
            }
            length += count;
        }

        int sequenceStart = -1;
        int position = 0;
        boolean lineStart = true;

        while (position < length) {
            byte current = data[position];

            if (lineStart && current == '>') {
                if (sequenceStart >= 0) {
                    reverseComplement(data, sequenceStart, position);
                }

                while (position < length
                        && data[position] != '\n'
                        && data[position] != '\r') {
                    position++;
                }
                if (position < length && data[position] == '\r') {
                    position++;
                }
                if (position < length && data[position] == '\n') {
                    position++;
                }

                sequenceStart = position;
                lineStart = true;
            } else {
                lineStart = current == '\n' || current == '\r';
                position++;
            }
        }

        if (sequenceStart >= 0) {
            reverseComplement(data, sequenceStart, length);
        }

        OutputStream output = System.out;
        output.write(data, 0, length);
        output.flush();
    }
}